Add the Generated Connector Code and Documentation (#2)
ballerina-platform/module-ballerinax-hubspot.automation.actions: README.md

## Overview

[HubSpot ](https://www.hubspot.com/) is an AI-powered customer platform with all the software, integrations, and resources you need to connect your marketing, sales, and customer service. HubSpot's connected platform enables you to grow your business faster by focusing on what matters most: your customers.

> [//]: # (TODO: Add overview mentioning the purpose of the module, supported REST API versions, and other high-level details.)

Reviewer: We can remove this comment.

Author: removed

> The `hubspot.automation.actions` package offers APIs to connect and interact with [Hubpsot Automation action API](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions) endpoints.

Reviewer: If we have a specific version in the API level, better to mention it.

Author: done - https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/acbce5b6cc9aaec87ffb6965c63cb150edbe13ab

## Setup guide

> [//]: # (TODO: Add detailed steps to obtain credentials and configure the module.)

Reviewer: Remove this todo-comment (please check for other places too)

Author: done -https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/acbce5b6cc9aaec87ffb6965c63cb150edbe13ab

'curl --request POST \
  --url https://api.hubapi.com/oauth/v1/token \
  --header 'content-type: application/x-www-form-urlencoded' \
  --data 'grant_type=authorization_code&code=<code>&redirect_uri=http://localhost:9090&client_id=<client_id>&client_secret=<client_secret>'

If the command executes successfully , you will receive the access token from the response.

### Step 10 : Obtaining the developer API key .

Follow the instructions at (https://developers.hubspot.com/docs/api/developer-tools-overview#developer-api-keys) to obtain the developer API key.

## Quickstart 

To begin using the `Hubspot Automation API` connector in your Ballerina application, you'll need to follow these steps:

### Step 1: Import the connector

First, import the `ballerinax/hubspot.automation.actions` package into your Ballerina project.

```ballerina
import ballerinax/hubspot.automation.actions;
```

### Step 2: Instantiate a new connector

Create a `hubspot.automation.actions:ConnectionConfig` object with your domain and developer API token, and initialize the connector.

```ballerina

> hubspot.automation.actions:ConnectionConfig automationClient = {
>    auth:{
> 
>     hapikey:"<developer-api-key>" ,
>     private\-app\-legacy: ""
> }
> };
> 
> hubspot.automation.actions:Client automationClient = check new (automationClient);

Reviewer: ```suggestion
hubspot.automation.actions:ConnectionConfig config = {
   auth:{

    hapikey:"<developer-api-key>" ,
    private\-app\-legacy: ""
}
};

hubspot.automation.actions:Client hubspotAutomation = check new (config);
```

Author: updated -https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/4f6f77ab909304f734642c23e2725b4ebce11e6b

## Overview

[HubSpot](https://www.hubspot.com/) is an AI-powered customer relationship management (CRM) platform. 

> The `ballerinax/hubspot.automation.actions` offers APIs to connect and interact with the [Automation Actions](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)  endpoints, specifically based on the [API v4](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)

Reviewer 2: ```suggestion
The `ballerinax/hubspot.automation.actions` offers APIs to connect and interact with the [Automation Actions](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)  endpoints, specifically based on the [HubSpot REST API](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)
```

Reviewer 2: Update in other .md files too

> The `ballerinax/hubspot.automation.actions` offers APIs to connect and interact with the [Automation Actions](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)  endpoints, specifically based on the [API v4](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)

Reviewer 2: ```suggestion
The `ballerinax/hubspot.automation.actions` offers APIs to connect and interact with the [Automation Actions](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions) endpoints, specifically based on the [HubSpot REST API](https://developers.hubspot.com/docs/reference/api).
```

## Setup guide

To use the Hubspot Automation action API connector in Ballerina, you must have a HubSpot developer account. If you don't have an account, create one using the following steps.

### Step 1: Create a Hubspot Developer Account

Visit the [Hubspot portal](https://developers.hubspot.com/get-started) and create a Developer Account.

### Step 2: Create a Hubspot Developer Test Account

Visit [developer test account page](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) and create a Hubspot developer test account.

### Step 3: Create a Hubspot Public App

In your developer account, navigate to the "Apps" section.

Click on "Create App" and provide the necessary details, including the app name and description.
### Step 4: Initiate the OAuth Flow

> Move to the auth tab in the created app and set the permissions there.

Reviewer 2: ```suggestion
Move to the "Auth" tab in the created app and set the permissions there.
```

Go back to the Auth Page and add the required scopes under the Auth tab .

![Required Scopes](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/docs/setup/resources/scopes.png)

Save the app

![Save the app](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/docs/setup/resources/save.png)

### Step   7: Obtain the authorization code

Copy the App installation url and paste it in the web browser.

![Redirect URL ](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/docs/setup/resources/redirect.png)

 It wll prompt you to install  the App and then select your developer test account.

After selecting the developer test account, you will receive a authorization code displayed in the browser.

![Obtain the authorization code](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/docs/setup/resources/authorization_code.png)

### Step   8: Obtain the access token

> Place your **authorization_code**, **client_id** and **client_secret** in the folowing comand and execute it in the terminal

Reviewer 2: ```suggestion
Place your `<authorization_code>`, `<client_id>` and `<client_secret>` in the following command and execute it in the terminal
```

  --data 'grant_type=authorization_code&code=<authorization_code>&redirect_uri=http://localhost:9090&client_id=<client_id>&client_secret=<client_secret>'

In order to receive a token we need to run a listener at the provided redirect uri.

If the command executes successfully , you will receive the access token from the response.

### Step   9: Obtaining the developer API key

Follow the instructions at (https://developers.hubspot.com/docs/api/developer-tools-overview#developer-api-keys) to obtain the developer API key.

## Quickstart

To begin using the `Hubspot Automation API` connector in your Ballerina application, you'll need to follow these steps:

### Step 1: Import the connector

First, import the `ballerinax/hubspot.automation.actions` package into your Ballerina project.

```ballerina
import ballerinax/hubspot.automation.actions;
```

### Step 2: Instantiate a new connector

Create a `actions:ConnectionConfig` object with your domain and developer API token, and initialize the connector.

```ballerina
actions:ConnectionConfig config = {
   auth:{
    hapikey:"<developer-api-key>" ,
    private\-app\-legacy: ""
}

> };

Reviewer 2: ```suggestion
actions:ConnectionConfig config = {
   auth: {
    hapikey: "<developer-api-key>" ,
    private\-app\-legacy: ""
   }
};
```

> actions:Client hubspotAutomation = check new (config);

Reviewer 2: ```suggestion
final actions:Client hubspotAutomation = check new (config);
```

    functionType: "POST_ACTION_EXECUTION"
};

actions:PublicActionDefinitionEgg testingPublicActionDefinitionEgg = {
    inputFields: [inputFieldDefinition],
    actionUrl: "https://webhook.site/94d09471-6f4c-4a7f-bae2-c9a585dd41e0",
    published: false,
    objectTypes: ["CONTACT"],
    objectRequestOptions: { properties: ["email"] },
    functions: [publicActionFunction],
    labels: {
        "en": {
            "inputFieldLabels": {
                "staticInput": "Static Input",
                "objectInput": "Object Property Input",
                "optionsInput": "External Options Input"
            },
            "actionName": "My Extension",
            "actionDescription": "My Extension Description",
            "appDisplayName": "My App Display Name",
            "actionCardContent": "My Action Card Content"
        }
    }
};

actions: PublicActionDefinition response = check hubspotAutomation->/automation/v4/actions/[appId].post(testingPublicActionDefinitionEgg);
```

#### List definitions

```ballerina

actions : CollectionResponsePublicActionDefinitionForwardPaging response = check hubspotAutomation->/automation/v4/actions/[appId];

> 

Reviewer 2: ```suggestion
actions : CollectionResponsePublicActionDefinitionForwardPaging response = check hubspotAutomation->/automation/v4/actions/[appId];
```

```

## Examples

The `Hubspot Automation API` connector provides practical examples illustrating usage in various scenarios. Explore these [examples](https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/tree/main/examples/), covering the following use cases:

1. [Extension CRUD](https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/tree/main/examples/)- Perform CRUD operations on Extensions

> 2. [Call complete callback APIs](https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/tree/main/examples/) - complete callbacks using the  Hubspot API

Reviewer 2: ```suggestion
2. [Call complete callback APIs](https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/tree/main/examples/) - Complete callbacks using the  HubSpot API
```

## Overview

[HubSpot](https://www.hubspot.com/) is an AI-powered customer relationship management (CRM) platform. 

> The `ballerinax/hubspot.automation.actions` offers APIs to connect and interact with the [Automation Actions](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)  endpoints, specifically based on the [API v4](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)

Reviewer 3: ```suggestion
The `ballerinax/hubspot.automation.actions` offers APIs to connect and interact with the [Automation Actions](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions) endpoints, specifically based on the [HubSpot REST API](https://developers.hubspot.com/docs/reference/api/overview)
```

## Overview

[HubSpot ](https://www.hubspot.com/) is an AI-powered customer relationship management (CRM) platform. 

> The `ballerinax/hubspot.automation.actions` offers APIs to connect and interact with the [Automation Actions](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions)  endpoints , specifically based on the [API v4 documention](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions) 

Reviewer 3: ```suggestion
The `ballerinax/hubspot.automation.actions` offers APIs to connect and interact with the [Automation Actions](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions) endpoints, specifically based on the [API v4](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions). 
```

Author: done-https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/681c29299560f89fc55954d31252356fb160de84

> 

Reviewer 3: Please remove the extra empty lines.

```suggestion
```

> The `hubspot.automation.actions` package offers APIs to connect and interact with [Hubpsot Automation action API (v4)](https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions) endpoints.

Reviewer 3: We can remove this duplicated part.

```suggestion
```

## Setup guide

> To use the Hubspot Automation action API connector in Ballerina, you must have a HubSpot developer account.

Reviewer 3: ```suggestion
To use the Hubspot Automation action API connector in Ballerina, you must have a HubSpot developer account. If you don't have an account, create one using the following steps.
```

 ### Step 1: Create a Hubspot Developer Account

Visit the Hubspot portal (https://developers.hubspot.com/get-started) and create a Developer Account.

> 

Reviewer 3: Remove the extra empty lines. I'd suggest to use a markdown linter plugin in VS code (or any IDE you are using) to validate the MD files.

Author: done- https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/681c29299560f89fc55954d31252356fb160de84

>  ### Step 2: Create a Hubspot Developer Test Account

Reviewer 3: Don't use a space to start the line. Check the similar places.

```suggestion
### Step 2: Create a Hubspot Developer Test Account
```

Author: installed plugin and updated the docs -https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/681c29299560f89fc55954d31252356fb160de84

> Visit (https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts)and create a hubspot developer test account

Reviewer 3: ```suggestion
Visit [developer test account page](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) and create a Hubspot developer test account.
```

Author: done -https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/commit/fe92f88060ac298721d1efa1890f84012a9be895

>  ### Step 3: Create a Hubspot Public App

Reviewer 3: ```suggestion
### Step 3: Create a Hubspot Public App
```

Author: done -https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/681c29299560f89fc55954d31252356fb160de84

In your developer account, navigate to the "Apps" section.

Click on "Create App" and provide the necessary details, including the app name and description.

> 

Reviewer 3: ```suggestion
```

 ### Step 4: Initiate the OAuth Flow

> Move to the auth tab in the created app and set the permissions there .

Reviewer 3: ```suggestion
Move to the auth tab in the created app and set the permissions there.
```

> Under the OAuth tab you can find the following details ,

Reviewer 3: ```suggestion
Under the OAuth tab you can find the following details,
```

> 

Reviewer 3: ```suggestion
```

client_id: Your app's Client ID.

redirect_uri: The URL users will be redirected to after granting access.

> scope: A space-separated list of scopes your app is requesting.

Reviewer 3: Better to use bullet points here:
```suggestion
* `client_id`: Your app's Client ID.
* `redirect_uri`: The URL users will be redirected to after granting access.
* `scope`: A space-separated list of scopes your app is requesting.
```

> ![Auth Tab example](docs/setup/resources/auth.png)

Reviewer 3: Does this link work? I think we need to provide the relative path, which should be `../docs/...`. Can you please check in your forked repo whether this works?

```suggestion
![Auth Tab example](docs/setup/resources/auth.png)
```

Author: yes, it works in the forked repo

### Step   5: Add the redirect URL
Add your redirect url under the redirect urls.

> ![alt text](docs/setup/resources/redirect_url.png)

Reviewer 3: It would be better to add an alt text here.

```suggestion
![add the redirect url](../docs/setup/resources/redirect_url.png)
```

### Step   6: Add the Required Scopes

> Go to the relevant API documentation (https://developers.hubspot.com/docs/reference/api/automation/custom-workflow-actions) and find out the required scopes.

Reviewer 3: For a specific connector, the scopes are predefined. Let's add the specific scope here.

Reviewer 2: Seems like this comment is not resolved. List the scopes in a bullet list
Ex: 
    ```* automation ```

> 

Reviewer 3: ```suggestion
```

> ![Requied Scope doc](docs/setup/resources/scope_doc.png)

Reviewer 3: Please use a spelling and grammar checker plugin to validate the document.

```suggestion
![Required Scope doc](docs/setup/resources/scope_doc.png)
```

> 

Reviewer 3: ```suggestion
```

> Now come back to the Auth Page and add the required scopes under the Auth tab .

Reviewer 3: Let's not use the word `Now` in the steps.

```suggestion
Go back to the Auth Page and add the required scopes under the Auth tab.
```

> 

Reviewer 3: ```suggestion
```

![Required Scopes](docs/setup/resources/scopes.png)

> Now save the app 

Reviewer 3: It would be better if we can use bullet points in these sub steps as well.

```suggestion
Save the app
```

![Save the app](docs/setup/resources/save.png)

### Step   7: Obtain the authorization code 

Copy the App installation url and paste it in the web browser.

![Redirect URL ](docs/setup/resources/redirect.png)

 It wll prompt you to install  the App and then select your deveper test account.

> 

Reviewer 3: ```suggestion
```

After selcting the developertest account, you will receive a authorization code displalyed in the browser.

![Obtain the authorization code](docs/setup/resources/authorization_code.png)

### Step   8: Obtain the access token

> Place your authorization code, client_id and client_secret in the folowing comand and execute it in the terminal

Reviewer 3: ```suggestion
Place your `authorization_code`, `client_id` and `client_secret` in the folowing comand and execute it in the terminal
```

'curl --request POST \
  --url https://api.hubapi.com/oauth/v1/token \
  --header 'content-type: application/x-www-form-urlencoded' \

>   --data 'grant_type=authorization_code&code=<code>&redirect_uri=http://localhost:9090&client_id=<client_id>&client_secret=<client_secret>'

Reviewer 3: ```suggestion
  --data 'grant_type=authorization_code&code=<authorization code>&redirect_uri=http://localhost:9090&client_id=<client_id>&client_secret=<client_secret>'
```

> If the command executes successfully , you will receive the access token from the response.

Reviewer 3: ```suggestion
If the command executes successfully, you will receive the access token from the response.
```

> If the command executes successfully , you will receive the access token from the response.

Reviewer 3: To receive the token, you need a listener to run. Let's add that information here. We don't need the Ballerina code snippet here. Just mention that they need to run a listener locally to listen on port `9090` to receive the token.

Author: done -https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/681c29299560f89fc55954d31252356fb160de84

### Step   9: Obtaining the developer API key .

Follow the instructions at (https://developers.hubspot.com/docs/api/developer-tools-overview#developer-api-keys) to obtain the developer API key.

> 
> 

Reviewer 3: Remove extra empty lines, check all the places.

```suggestion
```

Author: done - https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/commit/681c29299560f89fc55954d31252356fb160de84

## Quickstart 

To begin using the `Hubspot Automation API` connector in your Ballerina application, you'll need to follow these steps:

### Step 1: Import the connector

First, import the `ballerinax/hubspot.automation.actions` package into your Ballerina project.

```ballerina
import ballerinax/hubspot.automation.actions;
```

### Step 2: Instantiate a new connector

Create a `actions:ConnectionConfig` object with your domain and developer API token, and initialize the connector.

```ballerina
actions:ConnectionConfig config = {
   auth:{
    hapikey:"<developer-api-key>" ,
    private\-app\-legacy: ""
}
};
actions:Client hubspotAutomation = check new (config);
```

### Step 3: Invoke the connector operation

Utilize the connector's operations to manage extentions and functions.

> #### Create an extenstion

Reviewer 3: ```suggestion

#### Create an extension
```

Author: done - https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/pull/2/commits/681c29299560f89fc55954d31252356fb160de84

    automationFieldType: "",
    typeDefinition: typeDefinition,
    supportedValueTypes: ["STATIC_VALUE"]
};

actions:PublicActionFunction publicActionFunction = {
    functionSource: "exports.main = (event, callback) => {\r\n  callback({\r\n    outputFields: {\r\n      myOutput: \"example output value\"\r\n    }\r\n  });\r\n}",
    functionType: "POST_ACTION_EXECUTION"
};

actions:PublicActionDefinitionEgg testingPublicActionDefinitionEgg = {
    inputFields: [inputFieldDefinition],
    actionUrl: "https://webhook.site/94d09471-6f4c-4a7f-bae2-c9a585dd41e0",
    published: false,
    objectTypes: ["CONTACT"],
    objectRequestOptions: { properties: ["email"] },
    functions: [publicActionFunction],
    labels: {
        "en": {
            "inputFieldLabels": {
                "staticInput": "Static Input",
                "objectInput": "Object Property Input",
                "optionsInput": "External Options Input"
            },
            "actionName": "My Extension",
            "actionDescription": "My Extension Description",
            "appDisplayName": "My App Display Name",
            "actionCardContent": "My Action Card Content"
        }
    }
};

> 
> 
> 

Reviewer 3: ```suggestion
```

actions: PublicActionDefinition response = check hubspotAutomation->/automation/v4/actions/[appId].post(testingPublicActionDefinitionEgg);
```

#### List definitions

```ballerina

actions : CollectionResponsePublicActionDefinitionForwardPaging response = check hubspotAutomation->/automation/v4/actions/[appId];

```

## Examples

The `Hubspot Automation API` connector provides practical examples illustrating usage in various scenarios. Explore these [examples](../examples/), covering the following use cases:

1. [Extension CRUD](../examples/) - Perform CRUD operations on Extensions
2. [Call complete callback APIs](../examples/) - complete callbacks using the  Hubspot API

> 
> 
> 
> 

Reviewer 3: ```suggestion
```

Author: done -https://github.com/ballerina-platform/module-ballerinax-hubspot.automation.actions/commit/681c29299560f89fc55954d31252356fb160de84
